[Android] Fix issue creating ColorStateLists from extension methods

jsuarezruiz opened the pull request:
### Description of Change

Fix issue with Android CheckBox colors.

Disabled
![Screenshot_1653642321](https://user-images.githubusercontent.com/6755973/170674845-9d70fb8e-6eb7-476f-a99c-f547f4f1d441.png)

Normal
![Screenshot_1653642324](https://user-images.githubusercontent.com/6755973/170674847-2726e27b-26db-40ee-b275-f153f5a52f66.png)

Before these changes, The CheckBox disabled color was the same as enabled.

jsuarezruiz committed:
Fix issue with Android CheckBox colors

PureWeen reviewed (changes requested):
I don't think this fix is correct. If a user wants to express different colors for different states then they have to use a VSM.  Forcing the disabled color to be its own color apart from the colors specified on the control makes it impossible for a user to style the checkbox when it's "disabled". See from this example here against this PR.

I feel like we need to detect if a user has attached a VSM or not. Which I think is how WinUI works. So if the user hasn't attached a VSM then we could make it work like this. I think we did something like that for button in XF. Not to say we should do that with this PR but I don't think this PR is correct given the current design of the system. The original bug might be an issue with the VSM. The default styles in our template should cause the color to change when the checkbox state changes

<img width="992" alt="image" src="https://user-images.githubusercontent.com/5375137/170720261-41e7c487-e629-4c9b-be26-1fb531068f2c.png">

Symbai commented:
> Forcing the disabled color to be its own color apart from the colors specified on the control makes it impossible for a user to style the checkbox when it's "disabled"

I second this. I am running into a lot of issues with trying to change colors because there are so many "hardcoded" colors in the controls. This is a huge problem when a developer, like me, works with a custom background color.

jsuarezruiz commented:
> I don't think this fix is correct. If a user wants to express different colors for different states then they have to use a VSM. Forcing the disabled color to be its own color apart from the colors specified on the control makes it impossible for a user to style the checkbox when it's "disabled". See from this example here against this PR.
> 
> I feel like we need to detect if a user has attached a VSM or not. Which I think is how WinUI works. So if the user hasn't attached a VSM then we could make it work like this. I think we did something like that for button in XF. Not to say we should do that with this PR but I don't think this PR is correct given the current design of the system. The original bug might be an issue with the VSM. The default styles in our template should cause the color to change when the checkbox state changes
> 
> <img alt="image" width="992" src="https://user-images.githubusercontent.com/5375137/170720261-41e7c487-e629-4c9b-be26-1fb531068f2c.png">

Mnn, should we continue with the changes detecting if use VS, or not?. At least, we must leave the changes in src/Core/src/Platform/Android/ColorStateListExtensions.cs to get that extensions working as expected.

jsuarezruiz committed:
Use the same color in all the CheckBox states

jsuarezruiz commented:
I have removed the association with issue #6727  and updated the title. This PR still have sense to fix issues in the extension methods used to create ColorStateLists.

@PureWeen Could you review it again?. Thanks.

mattleibow commented:
> Mnn, should we continue with the changes detecting if use VS, or not?. At least, we must leave the changes in src/Core/src/Platform/Android/ColorStateListExtensions.cs to get that extensions working as expected.

I see what you mean. If the user adds a checkbox and does nothing else, the color should be platform defaults. Then setting a background color would tell me to only set the enabled state as I would expect the OS to gray out the control by default when disabled.

Now if there is a VSM set, then I would expect that to take precedence.

This is the same issue for all controls as I think they all go gray.

jsuarezruiz commented:
Applied some changes to fix https://github.com/dotnet/maui/issues/6727
Currently, we use the Color property value in the normal and disabled states. So, no visual differences enabling or disabling the CheckBox. I include some changes to use the Color property when disabling the control but applying a small change in the opacity of the color. Have sense?

![fix-6727-droid](https://user-images.githubusercontent.com/6755973/172371617-1183fb1a-6619-40fe-a961-cb797d51e51b.gif)

jsuarezruiz committed:
Changes to fix #6727

Symbai commented:
> but applying a small change in the opacity of the color

That's a very common solution on WPF, much better than before although personally I would reduce the opacity even further from 0.75 to 0.6 or 0.5. For end user it should be very noticeably if a control is disabled or not and I got better experience with lower opacity values. Especially if you got several of these controls in a list and not just two on a blank page like in your GIF.

rmarinho reviewed (changes requested):
The problem is if on Disable state i pass a color you are always changing it to 0.75 opacity. So we are preventing the user to set correct color.
I think we can do it if no color is set (we are using the default color) , otherwise the disableColor =  color set by the user

jsuarezruiz committed:
Revert disabled color changes

PureWeen commented:
> > Mnn, should we continue with the changes detecting if use VS, or not?. At least, we must leave the changes in src/Core/src/Platform/Android/ColorStateListExtensions.cs to get that extensions working as expected.
> 
> I see what you mean. If the user adds a checkbox and does nothing else, the color should be platform defaults. Then setting a background color would tell me to only set the enabled state as I would expect the OS to gray out the control by default when disabled.
> 
> Now if there is a VSM set, then I would expect that to take precedence.
> 
> This is the same issue for all controls as I think they all go gray.

I don't think this is the same for all controls. For example on iOS and WinUI we overwrite every single state when you set a color

![image](https://user-images.githubusercontent.com/5375137/175351856-882acfd6-0d87-4be8-983f-137f3453d82a.png)

![image](https://user-images.githubusercontent.com/5375137/175351960-1b73eca4-10f2-439b-a308-bd3e1e647493.png)

This is where it'd be nice to have some sort of state based understanding in the interface layer. Without that, inside core we just have to blow away the color for all states because we have no context that the user wants a color set. 

If we want to go the WPF/WinUI route that's going to be a bigger change at the controls layer where we force users to use the VSM in order to set colors in all states except Normal. Which I think is how WinUI and those work.

PureWeen committed:
- set accent color from context

PureWeen commented on src/Core/src/Platform/Android/CheckBoxExtensions.cs:
```diff
@@ -23,28 +23,30 @@ public static void UpdateIsChecked(this AppCompatCheckBox platformCheckBox, IChe
 
-			CompoundButtonCompat.SetButtonTintList(platformCheckBox, tintList);
-			CompoundButtonCompat.SetButtonTintMode(platformCheckBox, tintMode);
+			return ColorStateListExtensions.CreateSwitch(tintColor, tintColor, tintColor);
```
Should this be CreateCheckbox?

PureWeen committed:
- fix extension and add samples

PureWeen committed:
- add checked

PureWeen committed:
- only pull accent color if you need to
